---
title: "shRNA design to target circRNA and linear RNA splicing sequences"
output: rmarkdown::html_vignette
html_document:
    toc: true
    toc_depth: 2
    number_sections: true
    df_print: paged
    theme: united
    toc_float: true
    collapsed: false
vignette: >
  %\VignetteIndexEntry{shRNA tools}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Instalation and Setup

```{r setup}
#install library
#devtools::install_github("ipatop/shRNAtools")
library(shRNAtools)
```

# Description

This package contains two functions that will generate a Data Frame or tab separated file with the original table plus the shRNA Oligo design appended. 

The input should be a tab separated table with circRNA or linear RNA splicing junction coordinates and gene names in the format: 

<table>

Name    | Chr   | Start    | End      | Strand
------  | ----- | -------- | -------- | -------
circMbl | chr2R	| 17275410 | 17276063 |    +

</table>

This table can have an optional column with the Strand

So far the following species are available: 

+ Fly: 
  + dm3
  + dm6 
+ Human: 
  + hs19
  + hg38
+ Mice: 
  + mm10 
  + mm39
+ Rat: 
  + rn4

# Run circRNA example

Input should look like this
```{r}
#Copy and paste the full path of the test data into the readtable argument
system.file("extdata", "circs_totest.txt", package = "shRNAtools")

read.table(file = "/Users/inespatop/Documents/shRNA_design_circRNA/shRNAtools/inst/extdata/circs_totest.txt",header = T)
```

Run to create an output table
```{r}
#Copy and paste the full path of the test data into the readtable argument
system.file("extdata", "circs_totest.txt", package = "shRNAtools")

circRNASpliceOligoDesigner(input_coordinates  = "../test/circs_totest.txt",output = "../test/New_out.tsv")
```

Run to create a DataFrame, if writetab = F
```{r}
circOligos<-circRNASpliceOligoDesigner(input_coordinates = "../test/circs_totest.txt",output = "../test/New_out.tsv",writetab = F)
```

**Output**
```{r}
as.data.frame(circOligos)
```

# Design shRNA agains linear RNA splice junction

Input should look like this
```{r}
#Copy and paste the full path of the test data into the readtable argument
system.file("extdata", "lin_totest.txt", package = "shRNAtools")

read.table(file = "/Users/inespatop/Documents/shRNA_design_circRNA/shRNAtools/inst/extdata/lin_totest.txt",header = T)
```

Run to create an output table
```{r}
#Copy and paste the full path of the test data into the readtable argument
system.file("extdata", "lin_totest.txt", package = "shRNAtools")

linRNASpliceOligoDesigner(input_coordinates  = "../test/circs_totest.txt",output = "../test/New_out_lin.tsv")
```

Run to create a DataFrame, if writetab = F
```{r}
linOligos<-linRNASpliceOligoDesigner(input_coordinates = "../test/circs_totest.txt",output = "../test/New_out.tsv",writetab = F)
```

**Output**
```{r}
as.data.frame(linOligos)
```

# Find RCMs (reverse complementary repeats) in introns flanking circRNAs

Input should look like this
```{r}
#Copy and paste the full path of the test data into the readtable argument
system.file("extdata", "introns.txt", package = "shRNAtools")

read.table(file = "/Users/inespatop/Documents/shRNA_design_circRNA/shRNAtools/inst/extdata/introns.txt",header = T)
```

Run to create an output table
```{r}
RCMs <- find_rcm(sp = "mm39",intron_coordinates  = "/Users/inespatop/Documents/shRNA_design_circRNA/shRNAtools/inst/extdata/introns.txt",outBlast = "outBlast.tsv",outBlast.text = "outBlast.txt",blastn = "/Users/inespatop/Documents/Scripts/blast-2.14.0-h23b05c9_0/bin/blastn",ret = T)
```

**Output**
Table
```{r}
as.data.frame(RCMs)
```

# Requiered packages

The following packages are required, here an example of how to install it

```
install.packages("dplyr")
install.packages("tidyr")
BiocManager::install("Rsubread")
BiocManager::install("GenomicFeatures")
BiocManager::install("Biostrings")
BiocManager::install("TxDb.Dmelanogaster.UCSC.dm6.ensGene")
BiocManager::install("BSgenome.Dmelanogaster.UCSC.dm6")
BiocManager::install("TxDb.Dmelanogaster.UCSC.dm3.ensGene")
BiocManager::install("BSgenome.Dmelanogaster.UCSC.dm3")
BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
BiocManager::install("TxDb.Mmusculus.UCSC.mm10.knownGene")
BiocManager::install("BSgenome.Mmusculus.UCSC.mm10")
BiocManager::install("TxDb.Rnorvegicus.UCSC.rn4.ensGene")
BiocManager::install("BSgenome.Rnorvegicus.UCSC.rn4")
BiocManager::install("TxDb.Mmusculus.UCSC.mm39.refGene")
```


